### ISSUE TITLE
react-native run-android error “Task :app:javaPreCompileDebug FAILED”

### ISSUE THREAD

**Issue by GPaoloni** [facebook/react-native#25824]
I need to use some native code in a react native project. The native code is provided in a .jar file by the hardware manufacturer. In a clean project (react-native init), with this .jar file inside /android/app/libs folder, this error shows as soon as I try to run "react native run-android".

> Task :app:javaPreCompileDebug FAILED 6 actionable tasks: 6 executed
> 
> FAILURE: Build failed with an exception. What went wrong: Execution failed for task ':app:javaPreCompileDebug'. Could not resolve all files for configuration ':app:debugCompileClasspath'. Failed to transform file 'iMagPayV5.1.4.jar' to match attributes {artifactType=android-classes, org.gradle.usage=java-runtime-jars} Execution failed for JetifyTransform: /home/buonapasta/Desktop/React-Native/samples/project/android/app/libs/iMagPayV5.1.4.jar. Failed to transform '/home/buonapasta/Desktop/React-Native/samples/project/android/app/libs/iMagPayV5.1.4.jar' using Jetifier. Reason: 10. (Run with --stacktrace for more details.)
> 

Its worth to mention that we are already using the .jar library in an Android Studio project and it is working OK. Also I've already tried including some other .jar in my react native project and that is working OK too.

React Native version:
System:
    OS: Linux 4.18 Ubuntu 18.04.2 LTS (Bionic Beaver)
    CPU: (4) x64 Intel(R) Core(TM) i5-4300M CPU @ 2.60GHz
    Memory: 2.86 GB / 7.67 GB
    Shell: 4.4.20 - /bin/bash
  Binaries:
    Node: 10.16.0 - /usr/bin/node
    Yarn: 1.17.3 - /usr/bin/yarn
    npm: 6.10.2 - /usr/bin/npm
  npmPackages:
    react: 16.8.6 => 16.8.6 
    react-native: 0.60.4 => 0.60.4 
  npmGlobalPackages:
    react-native-cli: 2.0.1
    react-native-expo-svg: 1.0.1

## Steps To Reproduce

1. react-native init project
2. cd project && react-native run-android (to check it is working)
3. include this .jar in project/android/app/libs (creating the folder if it does not exist, acording to official docs)
4. react-native run-android (to try installing the app with the .jar)

Describe what you expected to happen:
Being able to include the .jar library to be used in native (android) code.

Snack, code example, screenshot, or link to a repository:

android/app/build.gradle looks like this:
```
... 

dependencies {
    implementation fileTree(dir: "libs", include: ["*.jar"])
    implementation "com.facebook.react:react-native:+"  // From node_modules

    if (enableHermes) {
      def hermesPath = "../../node_modules/hermesvm/android/";
      debugImplementation files(hermesPath + "hermes-debug.aar")
      releaseImplementation files(hermesPath + "hermes-release.aar")
    } else {
      implementation jscFlavor
    }
}

// Run this once to be able to run the application with BUCK
// puts all compile dependencies into folder libs for BUCK to use
task copyDownloadableDepsToLibs(type: Copy) {
    from configurations.compile
    into 'libs'
}


```
android/build.gradle:

```
// Top-level build file where you can add configuration options common to all sub-projects/modules.

buildscript {
    ext {
        buildToolsVersion = "28.0.3"
        minSdkVersion = 16
        compileSdkVersion = 28
        targetSdkVersion = 28
        supportLibVersion = "28.0.0"
    }
    repositories {
        google()
        jcenter()
    }
    dependencies {
        classpath("com.android.tools.build:gradle:3.4.1")

        // NOTE: Do not place your application dependencies here; they belong
        // in the individual module build.gradle files
    }
}

allprojects {
    repositories {
        mavenLocal()
        maven {
            // All of React Native (JS, Obj-C sources, Android binaries) is installed from npm
            url("$rootDir/../node_modules/react-native/android")
        }
        maven {
            // Android JSC is installed from npm
            url("$rootDir/../node_modules/jsc-android/dist")
        }

        google()
        jcenter()
    }
}
```
I've tried doing like sugested in StackOverflow:
```
npm install --save-dev jetifier
npx jetify
```
But the error still the same.

Reading the npm page of jetifier, in the "Usage for jar/zip/aar files" section:
```
npm install jetifier
npx jetifier-standalone <your arguments here>
```
Runing this gives me the following error:
```
Exception in thread "main" java.lang.ArrayIndexOutOfBoundsException: 10
    at org.objectweb.asm.ClassReader.readFrameType(ClassReader.java:2313)
    at org.objectweb.asm.ClassReader.readFrame(ClassReader.java:2269)
    at org.objectweb.asm.ClassReader.readCode(ClassReader.java:1448)
    at org.objectweb.asm.ClassReader.readMethod(ClassReader.java:1126)
    at org.objectweb.asm.ClassReader.accept(ClassReader.java:698)
    at org.objectweb.asm.ClassReader.accept(ClassReader.java:500)
    at com.android.tools.build.jetifier.processor.transform.bytecode.ByteCodeTransformer.runTransform(ByteCodeTransformer.kt:39)
    at com.android.tools.build.jetifier.processor.Processor.visit(Processor.kt:443)
    at com.android.tools.build.jetifier.processor.archive.ArchiveFile.accept(ArchiveFile.kt:49)
    at com.android.tools.build.jetifier.processor.Processor.visit(Processor.kt:425)
    at com.android.tools.build.jetifier.processor.archive.Archive.accept(Archive.kt:76)
    at com.android.tools.build.jetifier.processor.Processor.transformLibrary(Processor.kt:421)
    at com.android.tools.build.jetifier.processor.Processor.transform(Processor.kt:247)
    at com.android.tools.build.jetifier.processor.Processor.transform$default(Processor.kt:234)
    at com.android.tools.build.jetifier.standalone.Main.run(Main.kt:157)
    at com.android.tools.build.jetifier.standalone.Main$Companion.main(Main.kt:109)
    at com.android.tools.build.jetifier.standalone.Main.main(Main.kt)
```

**Comment by abdullateef97** [comment]
hey @GPaoloni you found a fix yet

**Comment by GPaoloni** [comment]
@abdullateef97 no, not yet. However I found that the Java library makes use of some "deprecated" imports (BasicNameValuePair from apache). Can that be the cause?

**Comment by GPaoloni** [comment]
Forgot to close this issue, as this was already solved.
Apparently, there are some cases where jetify can't transform the .jar library (because of deprecated includes, but if the lib is from a 3rd party there is no workaround). I found 2 solutions:
1- disable jetify from gradle.properties
`android.useAndroidX=true
android.enableJetifier=true`
2- this magic line ("experimental and unsoported", but completely functional at least to us) that prevents jetify from transforming the specified libs
`android.jetifier.blacklist = your-troublemaker-library.jar`

Closing the issue. Hope this helps anybody struggling with the same scenario :)
